test_rmw_implementation/test/test_publisher.cpp

```diff
@@ -0,0 +1,245 @@
+  public CLASSNAME (TestPublisher, RMW_IMPLEMENTATION)
+{
+protected:
+  using Base = CLASSNAME (TestPublisher, RMW_IMPLEMENTATION);
+
+  void SetUp() override
+  {
+    Base::SetUp();
+    rmw_publisher_options_t options = rmw_get_default_publisher_options();
+    pub = rmw_create_publisher(node, ts, topic_name, qos_profile, &options);
+    ASSERT_NE(nullptr, pub) << rmw_get_error_string().str;
+  }
+
+  void TearDown() override
+  {
+    rmw_ret_t ret = rmw_destroy_publisher(node, pub);
+    EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+    Base::TearDown();
+  }
+
+  rmw_publisher_t * pub;
+  const char * const topic_name{"/test"};
+  const rosidl_message_type_support_t * ts{
+    ROSIDL_GET_MSG_TYPE_SUPPORT(test_msgs, msg, BasicTypes)};
+  const rmw_qos_profile_t * qos_profile{&rmw_qos_profile_default};
+};
+
+TEST_F(CLASSNAME(TestPublisher, RMW_IMPLEMENTATION), create_and_destroy) {
+  rmw_publisher_options_t options = rmw_get_default_publisher_options();
+  const char * const topic_name = "/test";
+  const rosidl_message_type_support_t * ts =
+    ROSIDL_GET_MSG_TYPE_SUPPORT(test_msgs, msg, BasicTypes);
+  rmw_publisher_t * pub =
+    rmw_create_publisher(node, ts, topic_name, &rmw_qos_profile_default, &options);
+  ASSERT_NE(nullptr, pub) << rmw_get_error_string().str;
+  rmw_ret_t ret = rmw_destroy_publisher(node, pub);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+
+  const char * const native_topic_name = "test";
+  rmw_qos_profile_t native_qos_profile = rmw_qos_profile_default;
+  native_qos_profile.avoid_ros_namespace_conventions = true;
+  pub = rmw_create_publisher(node, ts, native_topic_name, &native_qos_profile, &options);
+  ASSERT_NE(nullptr, pub) << rmw_get_error_string().str;
+  ret = rmw_destroy_publisher(node, pub);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+}
+
+TEST_F(CLASSNAME(TestPublisher, RMW_IMPLEMENTATION), create_with_bad_arguments) {
+  rmw_publisher_options_t options = rmw_get_default_publisher_options();
+  const char * const topic_name = "/test";
+  const rosidl_message_type_support_t * ts =
+    ROSIDL_GET_MSG_TYPE_SUPPORT(test_msgs, msg, BasicTypes);
+  rmw_publisher_t * pub =
+    rmw_create_publisher(nullptr, ts, topic_name, &rmw_qos_profile_default, &options);
+  EXPECT_EQ(nullptr, pub);
+  rmw_reset_error();
+
+  pub = rmw_create_publisher(node, nullptr, topic_name, &rmw_qos_profile_default, &options);
+  EXPECT_EQ(nullptr, pub);
+  rmw_reset_error();
+
+  const char * implementation_identifier = node->implementation_identifier;
+  node->implementation_identifier = "not-an-rmw-implementation-identifier";
+  pub = rmw_create_publisher(node, ts, topic_name, &rmw_qos_profile_default, &options);
+  node->implementation_identifier = implementation_identifier;
+  EXPECT_EQ(nullptr, pub);
+  rmw_reset_error();
+
+  pub = rmw_create_publisher(node, ts, nullptr, &rmw_qos_profile_default, &options);
+  ASSERT_EQ(nullptr, pub);
```

Comment:
IMO those tests are independent, and if one fails doesn't affect the others, so the ASSERT statements can be changed to EXPECT ones.
```suggestion
  EXPECT_EQ(nullptr, pub);
```

Reply:
Good point. Also, there's a test above that should be two separate tests.